### DISCUSSION TITLE
Default Value Operator

### DISCUSSION THREAD

**Discussion by MichaelPuckett2** [dotnet/csharplang#2087] [Language Ideas]
Allow types to declare default values in the same way we use custom operators today.

Accessing the default value could be something now exposed via property but it could also be a shortcut to an existing property.

Using ```Nullable<T>``` for example but not only for Nullable<T>

Currently you can compare or assign with custom operators but you still have to access ```Value``` directly. 

Using a ```Nullable<T>``` for example:

Currently looks a little like this:

```
TimeSpan? ts1 = new TimeSpan(0, 30, 0);

if (ts1.HasValue) 
{
     Console.WriteLine($"{ts1.Value.Hours}:{ts1.Value.Minutes}:{ts1.Value.Seconds}");
}
```

Operator might look something like this:

```
public static default T(Nullable<T> nullable)
{
    return nullable.Value;
}
```

Now you could access ```Value``` with a double dot operator making it easier to read and a bit more common for other types doing the same thing.

```
TimeSpan? ts1 = new TimeSpan(0, 30, 0);

if (ts1.HasValue) 
{
     Console.WriteLine($"{ts1..Hours}:{ts1..Minutes}:{ts1..Seconds}");
}
```

**Comment by CyrusNajmabadi** [comment]
Right now the LDM is considering just allowing you to write things like `ts1.Hours` and having that be lifted automatically.  So this would be superfluous with that approach.

**Comment by MichaelPuckett2** [comment]
Can you send the link to those details?  I'm curious how this is going to be achieved; which I believe is a good idea; but if it's only for ```Nullable<T>``` then I feel like it's not getting us far.

**Comment by CyrusNajmabadi** [comment]
>  but if it's only for Nullable<T> then I feel like it's not getting us far.

What other scenarios are do you need this for?  Can you update the OP with those examples?

**Comment by yaakov-h** [comment]
I assume you're referring to #1865?

**Comment by MichaelPuckett2** [comment]
@CyrusNajmabadi  I am the OP... but yes I'll update the proposal with others.  I have mocked up examples but didn't want to clutter it up.  Having ```Nullable<T>``` return the inner type by default was not the goal... The goal was to allow any ```Type``` to have the ability to add a default value operator.

**Comment by MichaelPuckett2** [comment]
@yaakov-h No, that's not the proposal.  Maybe I shouldn't have used ```Nullable<T>``` for my example but it's the most obvious and talked about; it satisfies that situation as well.

**Comment by dsaf** [comment]
Would that work with ```struct```s?

**Comment by orthoxerox** [comment]
@dsaf structs are less of a problem, since separate specialized methods are compiled for struct generic parameters. I'm more worried about `default(T)` on a reference type, since it get's compiled into a null for all reference types.

**Comment by Thaina** [comment]
@CyrusNajmabadi 

> Right now the LDM is considering just allowing you to write things like `ts1.Hours` and having that be lifted automatically. So this would be superfluous with that approach.

Is it possible to did that lifting with `Task<T>` ?

**Comment by Thaina** [comment]
@MichaelPuckett2 Actually I think what you want is #889
